File: decision_tree.py
################### problem1#################################################
import pandas as pd
import numpy as np

#load the dataset
comp_data = pd.read_csv("C:/Users/DELL/Downloads/Decision Tree/Company_Data.csv")

comp_data.isnull().sum()
comp_data.dropna()
comp_data.columns
comp_data = comp_data.drop(["Age"], axis = 1)
comp_data['Sales'].describe()

#data pre-processing
#creat dummies for comp_data
comp_data = pd.get_dummies(comp_data, columns = ["ShelveLoc","Urban","US"])

#converting continous type to categorical
max = comp_data['Sales'].max()
comp_data['Sales'] = pd.cut(comp_data.Sales, bins = [-999 , max/2 , 999] , labels=['low' , 'high'])

comp_data['Sales'].unique()
comp_data['Sales'].value_counts()
colnames = list(comp_data.columns)

predictors = colnames[1:]
target = colnames[0]

# Splitting data into training and testing data set
from sklearn.model_selection import train_test_split
train, test = train_test_split(comp_data, test_size = 0.3)

from sklearn.tree import DecisionTreeClassifier as DT

help(DT)
model = DT(criterion = 'entropy')
model.fit(train[predictors], train[target])

# Prediction on Test Data
preds = model.predict(test[predictors])
pd.crosstab(test[target], preds, rownames=['Actual'], colnames=['Predictions'])

np.mean(preds == test[target]) # Test Data Accuracy 

# Prediction on Train Data
preds = model.predict(train[predictors])
pd.crosstab(train[target], preds, rownames = ['Actual'], colnames = ['Predictions'])

np.mean(preds == train[target]) # Train Data Accuracy

#overfitting issues occurs test accuracy is high and training accuracy is low
#using pruning techniques to overcome this issues

import pandas as pd
import numpy as np

#load the dataset
comp_data = pd.read_csv("C:/Users/DELL/Downloads/Decision Tree/Company_Data.csv")
#Drop the age coloumn
comp_data = comp_data.drop(["Age"], axis = 1)

#data pre-processing
#creat dummies for comp_data
comp_data = pd.get_dummies(comp_data, columns = ["ShelveLoc","Urban","US"])

comp_data['Sales'].unique()
comp_data['Sales'].value_counts()
colnames = list(comp_data.columns)

predictors = colnames[1:]
target = colnames[0]

# Splitting data into training and testing data set
from sklearn.model_selection import train_test_split
train, test = train_test_split(comp_data, test_size = 0.3)

# Train the Regression DT
from sklearn import tree
regtree = tree.DecisionTreeRegressor(max_depth = 5,ccp_alpha= 0.05)
regtree.fit(train[predictors], train[target])

# Prediction
test_pred = regtree.predict(test[predictors])
train_pred = regtree.predict(train[predictors])

# Measuring accuracy
from sklearn.metrics import mean_squared_error, r2_score

# Error on test dataset
mean_squared_error(test[target], test_pred)
r2_score(test[target], test_pred)

# Error on train dataset
mean_squared_error(train[target], train_pred)
r2_score(train[target], train_pred)

# Minimum observations at the internal node approach
regtree2 = tree.DecisionTreeRegressor(min_samples_split = 3)
regtree2.fit(train[predictors],train[target])

# Prediction
test_pred2 = regtree2.predict(test[predictors])
train_pred2 = regtree2.predict(train[predictors])

# Error on test dataset
mean_squared_error(test[target], test_pred2)
r2_score(test[target], test_pred2)

# Error on train dataset
mean_squared_error(train[target], train_pred2)
r2_score(train[target], train_pred2)

## Minimum observations at the leaf node approach
regtree3 = tree.DecisionTreeRegressor(min_samples_leaf = 3)
regtree3.fit(train[predictors], train[target])

# Prediction
test_pred3 = regtree3.predict(test[predictors])
train_pred3 = regtree3.predict(train[predictors])

# measure of error on test dataset
mean_squared_error(test[target], test_pred3)
r2_score(test[target], test_pred3)

# measure of error on train dataset
mean_squared_error(train[target], train_pred3)
r2_score(train[target], train_pred3)

###################################### random forest #################################
import pandas as pd
import numpy as np

#load the dataset
comp_data = pd.read_csv("C:/Users/DELL/Downloads/Decision Tree/Company_Data.csv")

comp_data.isnull().sum()
comp_data.dropna()
comp_data.columns
comp_data = comp_data.drop(["Age"], axis = 1)
comp_data['Sales'].describe()

#data pre-processing
#creat dummies for comp_data
comp_data = pd.get_dummies(comp_data, columns = ["ShelveLoc","Urban","US"])

#Categories the continous data of Sales column
max = comp_data['Sales'].max()
comp_data['Sales'] = pd.cut(comp_data.Sales, bins = [-999 , max/2 , 999] , labels=['low' , 'high'])

comp_data['Sales'].unique()
comp_data['Sales'].value_counts()
colnames = list(comp_data.columns)

predictors = colnames[1:]
target = colnames[0]

# Splitting data into training and testing data set
from sklearn.model_selection import train_test_split
train, test = train_test_split(comp_data, test_size = 0.3)

#building randomforest model
from sklearn.ensemble import RandomForestClassifier

rf_clf = RandomForestClassifier(n_estimators=500, n_jobs=1, random_state=42)

rf_clf.fit(train[predictors], train[target])

from sklearn.metrics import accuracy_score, confusion_matrix

#Test accuracy
confusion_matrix(test[target], rf_clf.predict(test[predictors]))
accuracy_score(test[target], rf_clf.predict(test[predictors]))

#Train accuracy
confusion_matrix(train[target], rf_clf.predict(train[predictors]))
accuracy_score(train[target], rf_clf.predict(train[predictors]))

#now its a right fit model

#################################### problem2 ###############################
#load the dataset
import pandas as pd
import numpy as np

#load the dataset
diabetes_data = pd.read_csv("C:/Users/DELL/Downloads/Decision Tree/Diabetes.csv")

diabetes_data.isnull().sum()
diabetes_data.dropna()
diabetes_data.columns

colnames = list(diabetes_data.columns)

predictors = colnames[:8]
target = colnames[8]

# Splitting data into training and testing data set
from sklearn.model_selection import train_test_split
train, test = train_test_split(diabetes_data, test_size = 0.3)

from sklearn.tree import DecisionTreeClassifier as DT

help(DT)
model = DT(criterion = 'entropy')
model.fit(train[predictors], train[target])

# Prediction on Test Data
preds = model.predict(test[predictors])
pd.crosstab(test[target], preds, rownames=['Actual'], colnames=['Predictions'])

np.mean(preds == test[target]) # Test Data Accuracy 

# Prediction on Train Data
preds = model.predict(train[predictors])
pd.crosstab(train[target], preds, rownames = ['Actual'], colnames = ['Predictions'])

np.mean(preds == train[target]) # Train Data Accuracy

#overfitting issues occurs test accuracy is high and training accuracy is low
#using pruning techniques to overcome this issues

import pandas as pd
import numpy as np

#load the dataset
diabetes_data = pd.read_csv("C:/Users/DELL/Downloads/Decision Tree/Diabetes.csv")

#creating dummy for categorical data
diabetes_data = pd.get_dummies(diabetes_data, columns = [" Class variable"])

colnames = list(diabetes_data.columns)

predictors = colnames[:8]
target = colnames[8]

# Splitting data into training and testing data set
from sklearn.model_selection import train_test_split
train, test = train_test_split(diabetes_data, test_size = 0.3)

# Train the Regression DT
from sklearn import tree
regtree = tree.DecisionTreeRegressor(max_depth = 5,ccp_alpha= 0.05)
regtree.fit(train[predictors], train[target])

# Prediction
test_pred = regtree.predict(test[predictors])
train_pred = regtree.predict(train[predictors])

# Measuring accuracy
from sklearn.metrics import mean_squared_error, r2_score

# Error on test dataset
mean_squared_error(test[target], test_pred)
r2_score(test[target], test_pred)

# Error on train dataset
mean_squared_error(train[target], train_pred)
r2_score(train[target], train_pred)

# Minimum observations at the internal node approach
regtree2 = tree.DecisionTreeRegressor(min_samples_split = 3)
regtree2.fit(train[predictors],train[target])

# Prediction
test_pred2 = regtree2.predict(test[predictors])
train_pred2 = regtree2.predict(train[predictors])

# Error on test dataset
mean_squared_error(test[target], test_pred2)
r2_score(test[target], test_pred2)

# Error on train dataset
mean_squared_error(train[target], train_pred2)
r2_score(train[target], train_pred2)

## Minimum observations at the leaf node approach
regtree3 = tree.DecisionTreeRegressor(min_samples_leaf = 3)
regtree3.fit(train[predictors], train[target])

# Prediction
test_pred3 = regtree3.predict(test[predictors])
train_pred3 = regtree3.predict(train[predictors])

# measure of error on test dataset
mean_squared_error(test[target], test_pred3)
r2_score(test[target], test_pred3)

# measure of error on train dataset
mean_squared_error(train[target], train_pred3)
r2_score(train[target], train_pred3)

###################################### random forest #################################
import pandas as pd
import numpy as np

#load the dataset
diabetes_data = pd.read_csv("C:/Users/DELL/Downloads/Decision Tree/Diabetes.csv")

#creating dummy for categorical data
diabetes_data = pd.get_dummies(diabetes_data, columns = [" Class variable"])

colnames = list(diabetes_data.columns)

predictors = colnames[:8]
target = colnames[8]

# Splitting data into training and testing data set
from sklearn.model_selection import train_test_split
train, test = train_test_split(diabetes_data, test_size = 0.3)

#building randomforest model
from sklearn.ensemble import RandomForestClassifier

rf_clf = RandomForestClassifier(n_estimators=500, n_jobs=1, random_state=42)

rf_clf.fit(train[predictors], train[target])

from sklearn.metrics import accuracy_score, confusion_matrix

#Test accuracy
confusion_matrix(test[target], rf_clf.predict(test[predictors]))
accuracy_score(test[target], rf_clf.predict(test[predictors]))

#Train accuracy
confusion_matrix(train[target], rf_clf.predict(train[predictors]))
accuracy_score(train[target], rf_clf.predict(train[predictors]))

#now its a right fit model

#################################### problem3 #############################################
import pandas as pd
import numpy as np

#load the dataset
fraud_check = pd.read_csv("C:/Users/DELL/Downloads/Decision Tree/Fraud_check.csv")

fraud_check.isnull().sum()
fraud_check.dropna()
fraud_check.columns
fraud_check['Taxable.Income'].describe()

#data pre-processing
#creat dummies for comp_data
fraud_check = pd.get_dummies(fraud_check, columns = ["Undergrad", "Marital.Status", "Urban"])


#converting continous type to categorical
max = fraud_check['Taxable.Income'].max()
fraud_check['Taxable.Income'] = pd.cut(fraud_check['Taxable.Income'], bins = [-999 , 30000 , 99999] , labels=['Risky' , 'Good'])

fraud_check['Taxable.Income'].unique()
fraud_check['Taxable.Income'].value_counts()
colnames = list(fraud_check.columns)

predictors = colnames[1:]
target = colnames[0]

# Splitting data into training and testing data set
from sklearn.model_selection import train_test_split
train, test = train_test_split(fraud_check, test_size = 0.3)

from sklearn.tree import DecisionTreeClassifier as DT

help(DT)
model = DT(criterion = 'entropy')
model.fit(train[predictors], train[target])

# Prediction on Test Data
preds = model.predict(test[predictors])
pd.crosstab(test[target], preds, rownames=['Actual'], colnames=['Predictions'])

np.mean(preds == test[target]) # Test Data Accuracy 

# Prediction on Train Data
preds = model.predict(train[predictors])
pd.crosstab(train[target], preds, rownames = ['Actual'], colnames = ['Predictions'])

np.mean(preds == train[target]) # Train Data Accuracy

#overfitting issues occurs test accuracy is high and training accuracy is low
#using pruning techniques to overcome this issues

import pandas as pd
import numpy as np

#load the dataset
fraud_check = pd.read_csv("C:/Users/DELL/Downloads/Decision Tree/Fraud_check.csv")

#data pre-processing
#creat dummies for comp_data
fraud_check = pd.get_dummies(fraud_check, columns = ["Undergrad", "Marital.Status", "Urban"])

colnames = list(fraud_check.columns)

predictors = colnames[1:]
target = colnames[0]

# Splitting data into training and testing data set
from sklearn.model_selection import train_test_split
train, test = train_test_split(fraud_check, test_size = 0.3)

# Train the Regression DT
from sklearn import tree
regtree = tree.DecisionTreeRegressor(max_depth = 5,ccp_alpha= 0.05)
regtree.fit(train[predictors], train[target])

# Prediction
test_pred = regtree.predict(test[predictors])
train_pred = regtree.predict(train[predictors])

# Measuring accuracy
from sklearn.metrics import mean_squared_error, r2_score

# Error on test dataset
mean_squared_error(test[target], test_pred)
r2_score(test[target], test_pred)

# Error on train dataset
mean_squared_error(train[target], train_pred)
r2_score(train[target], train_pred)

# Minimum observations at the internal node approach
regtree2 = tree.DecisionTreeRegressor(min_samples_split = 3)
regtree2.fit(train[predictors],train[target])

# Prediction
test_pred2 = regtree2.predict(test[predictors])
train_pred2 = regtree2.predict(train[predictors])

# Error on test dataset
mean_squared_error(test[target], test_pred2)
r2_score(test[target], test_pred2)

# Error on train dataset
mean_squared_error(train[target], train_pred2)
r2_score(train[target], train_pred2)

## Minimum observations at the leaf node approach
regtree3 = tree.DecisionTreeRegressor(min_samples_leaf = 3)
regtree3.fit(train[predictors], train[target])

# Prediction
test_pred3 = regtree3.predict(test[predictors])
train_pred3 = regtree3.predict(train[predictors])

# measure of error on test dataset
mean_squared_error(test[target], test_pred3)
r2_score(test[target], test_pred3)

# measure of error on train dataset
mean_squared_error(train[target], train_pred3)
r2_score(train[target], train_pred3)
#now its a right fit model
###################################### random forest #################################
import pandas as pd
import numpy as np

#load the dataset
fraud_check = pd.read_csv("C:/Users/DELL/Downloads/Decision Tree/Fraud_check.csv")

fraud_check.isnull().sum()
fraud_check.dropna()
fraud_check.columns

#data pre-processing
#creat dummies for comp_data
fraud_check = pd.get_dummies(fraud_check, columns = ["Undergrad","Marital.Status","Urban"])

#Categories the continous data of Sales column
max = fraud_check['Taxable.Income'].max()
fraud_check['Taxable.Income'] = pd.cut(fraud_check['Taxable.Income'], bins = [-999 ,30000 , 9999999] , labels=['low' , 'high'])

fraud_check['Taxable.Income'].unique()
fraud_check['Taxable.Income'].value_counts()
colnames = list(fraud_check.columns)

predictors = colnames[1:]
target = colnames[0]

# Splitting data into training and testing data set
from sklearn.model_selection import train_test_split
train, test = train_test_split(fraud_check, test_size = 0.3)

#building randomforest model
from sklearn.ensemble import RandomForestClassifier

rf_clf = RandomForestClassifier(n_estimators=500, n_jobs=1, random_state=42)

rf_clf.fit(train[predictors], train[target])

from sklearn.metrics import accuracy_score, confusion_matrix

#Test accuracy
confusion_matrix(test[target], rf_clf.predict(test[predictors]))
accuracy_score(test[target], rf_clf.predict(test[predictors]))

#Train accuracy
confusion_matrix(train[target], rf_clf.predict(train[predictors]))
accuracy_score(train[target], rf_clf.predict(train[predictors]))

#now its a right fit model

############################ problem4 ########################################
########################################Problem 4###########################################
#load the data

import pandas as pd
import numpy as np
HR_data = pd.read_csv("C:/Users/DELL/Downloads/Decision Tree/HR_DT.csv")

#dummy values
HR_data = pd.get_dummies(HR_data, columns = ["Position of the employee"])

#random Forest technique

# Splitting data into training and testing data set
colnames = list(HR_data.columns)
target = colnames[1]
predictors = colnames[:1]+colnames[2:]

from sklearn.model_selection import train_test_split
train, test = train_test_split(HR_data, test_size = 0.3)

#building the random forest model
from sklearn.ensemble import RandomForestClassifier

rf_clf = RandomForestClassifier(n_estimators=500, n_jobs=1, random_state=50)

rf_clf.fit(train[predictors], train[target])

from sklearn.metrics import accuracy_score, confusion_matrix

#accuracy on test data
confusion_matrix(test[target], rf_clf.predict(test[predictors]))
accuracy_score(test[target], rf_clf.predict(test[predictors]))

#accuracy on train data
confusion_matrix(train[target], rf_clf.predict(train[predictors]))
accuracy_score(train[target], rf_clf.predict(train[predictors]))

#create a dataframe of customers claim details
#filling Region Manager as 1 anyways we convert it to dummy here by saving computation
customer_claim_list = [[1, 5.0 , 70000]]

customer_claim = pd.DataFrame(customer_claim_list , columns= ["Position of the employee" , "no of Years of Experience of employee", " monthly income of employee"])

#combining and concatenating 2 dataframes
df = [test , customer_claim]
test = pd.concat(df)

#filling all na by 0
test = test.fillna(0)

#predicting using testing data where customer claims present at last row of test
preds = rf_clf.predict(test[predictors])

#storing predicted values in seperate column
test["predicted salary"] = preds

#accessing predicted value
test.iloc[59,13]

# predicted as 67938 , candidate claimed 70000 which is almost close.
#it can be assumed that candidate is genuine

###########################################END#####################################

File: decision_tree.r
################### problem1 ################################3
#load the dataset
company_data <- read.csv("C:\\Users\\DELL\\Downloads\\Decision Tree\\Company_Data.csv",stringsAsFactors = TRUE)
#converting continous data of target to categorical data
company_data$Sales <- cut(company_data$Sales,breaks = c(-Inf, 5.4, 11,Inf),labels = c("low","middle","high"))

##Exploring and preparing the data ----
str(company_data)
summary(company_data$Sales)

# look at the class variable
table(company_data$Sales)

#splitting train and test data
set.seed(0)
split <- sample.split(company_data$Sales, SplitRatio = 0.8)
company_data_train <- subset(company_data, split == TRUE)
company_data_test <- subset(company_data, split == FALSE)

# check the proportion of class variable for train and test data
prop.table(table(company_data$Sales))
prop.table(table(company_data_train$Sales))
prop.table(table(company_data_test$Sales))

# Step 3: Training a model on the data
install.packages("C50")
library(C50)

company_data_model <- C5.0(company_data_train[, -1], company_data_train$Sales)

windows()
plot(company_data_model) 

# Display detailed information about the tree
summary(company_data_model)

# Step 4: Evaluating model performance
# Test data accuracy
test_res <- predict(company_data_model, company_data_test)
test_acc <- mean(company_data_test$Sales == test_res)
test_acc

# cross tabulation of predicted versus actual classes
library(gmodels)
CrossTable(company_data_test$Sales, test_res, dnn = c('actual default', 'predicted default'))

# On Training Dataset
train_res <- predict(company_data_model, company_data_train)
train_acc <- mean(company_data_train$Sales == train_res)
train_acc

table(company_data_train$Sales, train_res)
#overfitting is occured due to training accuracy is high and testing accuracy is low

#prune the decision tree

library(rpart)
model <- rpart(company_data_train$Sales ~ .,data = company_data_train,method = "class",control = rpart.control(cp = 0, maxdepth = 3))

# Plot Decision Tree
install.packages("rpart.plot")
library(rpart.plot)
rpart.plot(model, box.palette = "auto", digits = -3)

# Evaluating model performance
# Test data accuracy
test_res <- predict(model, company_data_test,type = "class")
test_acc <- mean(company_data_test$Sales == test_res)
test_acc

# cross tabulation of predicted versus actual classes
library(gmodels)
CrossTable(company_data_test$Sales, test_res, dnn = c('actual default', 'predicted default'))

# On Training Dataset
train_res <- predict(model, company_data_train,type = "class")
train_acc <- mean(company_data_train$Sales == train_res)
train_acc

table(company_data_train$Sales, train_res)
# Now its a right fit model because train accuracy and test accuracy almost similar

################### Random forest ######################################################
##Exploring and preparing the data ----
# install.packages("randomForest")
install.packages("randomForest")
library(randomForest)

rf <- randomForest(company_data_train$Sales ~ ., data = company_data_train,ntree = 50,maxnodes = 15,method  = "class")

#on test data
test_res <- predict(rf, company_data_test,type = "class")
test_acc <- mean(company_data_test$Sales == test_res)
test_acc

# cross tabulation of predicted versus actual classes
library(gmodels)
CrossTable(company_data_test$Sales, test_res, dnn = c('actual default', 'predicted default'))

# On Training Dataset
train_res <- predict(rf, company_data_train,type = "class")
train_acc <- mean(company_data_train$Sales == train_res)
train_acc

table(company_data_train$Sales, train_res)

############################### problem2 ##########################################################################
#load the dataset
diabetes <- read.csv("C://Users//DELL//Downloads//Decision Tree//Diabetes.csv",stringsAsFactors = TRUE)
##Exploring and preparing the data ----
str(diabetes)

# look at the class variable
table(diabetes$Class.variable)

#prepare tarining and testing data
library(caTools)
set.seed(0)
split <- sample.split(diabetes$Class.variable, SplitRatio = 0.8)
diabetes_train <- subset(diabetes, split == TRUE)
diabetes_test <- subset(diabetes, split == FALSE)

# check the proportion of class variable
prop.table(table(diabetes$Class.variable))
prop.table(table(diabetes_train$Class.variable))
prop.table(table(diabetes_test$Class.variable))

# Step 3: Training a model on the data
install.packages("C50")
library(C50)

diabetes_model <- C5.0(diabetes_train[, -9], diabetes_train$Class.variable)

windows()
plot(diabetes_model) 

# Display detailed information about the tree
summary(diabetes_model)

# Step 4: Evaluating model performance
# Test data accuracy
test_res <- predict(diabetes_model, diabetes_test)
test_acc <- mean(diabetes_test$Class.variable == test_res)
test_acc

# cross tabulation of predicted versus actual classes
library(gmodels)
CrossTable(diabetes_test$Class.variable, test_res, dnn = c('actual default', 'predicted default'))

# On Training Dataset
train_res <- predict(diabetes_model, diabetes_train)
train_acc <- mean(diabetes_train$Class.variable == train_res)
train_acc

table(diabetes_train$Class.variable, train_res)
# overfitting issues since testing accuracy is low and traning accuracy is high

#pruning the model
library(rpart)
model <- rpart(diabetes_train$Class.variable ~ ., data = diabetes_train,method = "class",control = rpart.control(cp = 0, maxdepth = 3))

# Plot Decision Tree
library(rpart.plot)
rpart.plot(model, box.palette = "auto", digits = 4)

# Measure the RMSE on Test data
test_pred <- predict(model, newdata = diabetes_test, type = "class")

# RMSE
accuracy1 <- sqrt(mean(diabetes_test$Class.variable == test_pred)^2)
accuracy1

# Measure the RMSE on Train data
train_pred <- predict(model, newdata = diabetes_train, type = "class")

# RMSE
accuracy_train <- sqrt(mean(diabetes_train$Class.variable == train_pred)^2)
accuracy_train

#Training and Testing accuracy is almost similar hence it is right fit model

######################### Random Forest ###############################################
# install.packages("randomForest")
library(randomForest)

rf <- randomForest(diabetes_train$Class.variable ~ .,ntree = 5, maxnodes = 4  ,data = diabetes_train)

test_rf_pred <- predict(rf, diabetes_test)

rmse_rf <- sqrt(mean(diabetes_test$Class.variable == test_rf_pred)^2)
rmse_rf

# Prediction for trained data result
train_rf_pred <- predict(rf, diabetes_train)

# RMSE on Train Data
train_rmse_rf <- sqrt(mean(diabetes_train$Class.variable == train_rf_pred)^2)
train_rmse_rf

#Training and testing accuracy is almost simlilar hence it is a right fit model

############################### problem3 #####################################
##Exploring and preparing the data ----
str(fraud_check)
summary(company_data$Sales)

# look at the class variable
table(fraud_check$Taxable.Income)

#splitting train and test data
set.seed(0)
split <- sample.split(fraud_check$Taxable.Income, SplitRatio = 0.8)
fraud_check_train <- subset(fraud_check, split == TRUE)
fraud_check_test <- subset(fraud_check, split == FALSE)

# check the proportion of class variable for train and test data
prop.table(table(fraud_check$Taxable.Income))
prop.table(table(fraud_check_train$Taxable.Income))
prop.table(table(fraud_check_test$Taxable.Income))

# Step 3: Training a model on the data
install.packages("C50")
library(C50)

fraud_check_model <- C5.0(fraud_check_train[, -3], fraud_check_train$Taxable.Income)

windows()
plot(fraud_check_model) 

# Display detailed information about the tree
summary(fraud_check_model)

# Step 4: Evaluating model performance
# Test data accuracy
test_res <- predict(fraud_check_model, fraud_check_test)
test_acc <- mean(fraud_check_test$Taxable.Income == test_res)
test_acc

# cross tabulation of predicted versus actual classes
library(gmodels)
CrossTable(fraud_check_test$Taxable.Income, test_res, dnn = c('actual default', 'predicted default'))

# On Training Dataset
train_res <- predict(fraud_check_model, fraud_check_train)
train_acc <- mean(fraud_check_train$Taxable.Income == train_res)
train_acc

table(fraud_check_train$Taxable.Income, train_res)
#overfitting is occured 
#prune the decision tree

library(rpart)
model <- rpart(fraud_check_train$Taxable.Income ~ .,data = fraud_check_train,method = "class",control = rpart.control(cp = 0, maxdepth = 3))

# Plot Decision Tree
install.packages("rpart.plot")
library(rpart.plot)
rpart.plot(model, box.palette = "auto", digits = 2)

# Evaluating model performance
# Test data accuracy
test_res <- predict(model, fraud_check_test,type = "class")
test_acc <- mean(fraud_check_test$Taxable.Income == test_res)
test_acc

# cross tabulation of predicted versus actual classes
library(gmodels)
CrossTable(fraud_check_test$Taxable.Income, test_res, dnn = c('actual default', 'predicted default'))

# On Training Dataset
train_res <- predict(model, fraud_check_train,type = "class")
train_acc <- mean(fraud_check_train$Taxable.Income == train_res)
train_acc

table(fraud_check_train$Taxable.Income, train_res)
# Now its a right fit model because train accuracy and test accuracy almost similar

################### Random forest ######################################################
##Exploring and preparing the data ----
# install.packages("randomForest")
install.packages("randomForest")
library(randomForest)

rf <- randomForest(c$Taxable.Income ~ ., data = fraud_check_train,ntree = 50,maxnodes = 15,method  = "class")

#on test data
test_res <- predict(rf, fraud_check_test,type = "class")
test_acc <- mean(fraud_check_test$Taxable.Income == test_res)
test_acc

# cross tabulation of predicted versus actual classes
library(gmodels)
CrossTable(fraud_check_test$Taxable.Income, test_res, dnn = c('actual default', 'predicted default'))

# On Training Dataset
train_res <- predict(rf, fraud_check_train,type = "class")
train_acc <- mean(fraud_check_train$Taxable.Income == train_res)
train_acc

table(fraud_check_train$Taxable.Income, train_res)
#now model is right fit .

########################### problem4 #################################################
#load the dataset 
HR_data <- read.csv("C:\\Users\\DELL\\Downloads\\Decision Tree\\HR_DT.csv",stringsAsFactors = TRUE)

# look at the class variable
table(HR_data$monthly.income.of.employee)
str(HR_data)

summary(HR_data$monthly.income.of.employee)

#splitting train and test data
library(caTools)
set.seed(0)
split <- sample.split(HR_data$monthly.income.of.employee, SplitRatio = 0.8)
HR_data_train <- subset(HR_data, split == TRUE)
HR_data_test <- subset(HR_data, split == FALSE)

library(rpart)
model <- rpart(HR_data_train$monthly.income.of.employee ~ ., data = HR_data_train,method = "class",
               control = rpart.control(cp = 0, maxdepth = 5))

# Plot Decision Tree
library(rpart.plot)
rpart.plot(model, box.palette = "auto", digits = 3)


# Measure the RMSE on Test data
test_pred <- predict(model, newdata = HR_data_test, type = "vector")

# RMSE
accuracy1 <- sqrt(mean(HR_data_test$monthly.income.of.employee - test_pred)^2)
accuracy1

# Measure the RMSE on Train data
train_pred <- predict(model, newdata = HR_data_train, type = "vector")

# RMSE
accuracy_train <- sqrt(mean(HR_data_train$monthly.income.of.employee - train_pred)^2)
accuracy_train

# Prune the Decision Tree

# Grow the full tree
fullmodel <- rpart(HR_data_train$monthly.income.of.employee ~ ., data = HR_data_train,
                   control = rpart.control(cp = 0))

rpart.plot(fullmodel, box.palette = "auto", digits = 5)

# Examine the complexity plot
# Tunning parameter check the value of cp which is giving us minimum cross validation error (xerror)
printcp(fullmodel)   
plotcp(model)

mincp <- model$cptable[which.min(model$cptable[, "xerror"]), "CP"]

# Prune the model based on the optimal cp value
model_pruned_1 <- prune(fullmodel, cp = mincp)
rpart.plot(model_pruned_1, box.palette = "auto", digits = 5)

model_pruned_2 <- prune(fullmodel, cp = 0.02)
rpart.plot(model_pruned_2, box.palette = "auto", digits = 5)


# Measure the RMSE using Full tree
test_pred_fultree <- predict(fullmodel, newdata = HR_data_test, type = "vector")
# RMSE
accuracy_f <- sqrt(mean(HR_data_test$monthly.income.of.employee - test_pred_fultree)^2)
accuracy_f

# Measure the RMSE using Prune tree - model1
test_pred_prune1 <- predict(model_pruned_1, newdata = HR_data_test, type = "vector")
# RMSE
accuracy_prune1 <- sqrt(mean(HR_data_test$monthly.income.of.employee - test_pred_prune1)^2)
accuracy_prune1

# Measure the RMSE using Prune tree - model2
test_pred_prune2 <- predict(model_pruned_2, newdata = HR_data_test, type = "vector")
# RMSE
accuracy_prune2 <- sqrt(mean(HR_data_test$monthly.income.of.employee - test_pred_prune2)^2)
accuracy_prune2

# Prediction for trained data result
train_pred_fultree <- predict(fullmodel, HR_data_train, type = 'vector')

# RMSE on Train Data
train_accuracy_fultree <- sqrt(mean(HR_data_train$monthly.income.of.employee - train_pred_fultree)^2)
train_accuracy_fultree


# Prediction for trained data result
train_pred_prune1 <- predict(model_pruned_1, HR_data_train, type = 'vector')

# RMSE on Train Data
train_accuracy_fultree2 <- sqrt(mean(HR_data_train$monthly.income.of.employee - train_pred_prune1)^2)
train_accuracy_fultree2

# Prediction for trained data result
train_pred_prune2 <- predict(model_pruned_2, HR_data_train, type = 'vector')

# RMSE on Train Data
train_accuracy_fultree2 <- sqrt(mean(HR_data_train$monthly.income.of.employee - train_pred_prune2)^2)
train_accuracy_fultree2

#RMSE value for both test and accuracy is almost right hence its a right fit model

################# Random Forest #####################################################
# install.packages("randomForest")
library(randomForest)

rf <- randomForest(monthly.income.of.employee ~ .,maxnodes = 4,ntree = 3, data = HR_data_train)

test_rf_pred <- predict(rf, HR_data_test)

rmse_rf <- sqrt(mean(HR_data_test$monthly.income.of.employee - test_rf_pred)^2)
rmse_rf

# Prediction for trained data result
train_rf_pred <- predict(rf, HR_data_train)

# RMSE on Train Data
train_rmse_rf <- sqrt(mean(HR_data_train$monthly.income.of.employee - train_rf_pred)^2)
train_rmse_rf

#now checking that the candidate claim is genuine or fake
#candidate claims are stored in a dataframe
cols <- c("Position.of.the.employee", "no.of.Years.of.Experience.of.employee","monthly.income.of.employee")
Candidate_claim <- data.frame(a <-  "Region Manager" , b <-  5 , c <-  70000)
colnames(Candidate_claim) <- cols
#binding with test data for prediction
HR_data_test <- rbind(HR_data_test, Candidate_claim)

#predicting using model
HR_data_test$monthly.income.of.employee.pred <- predict(model, newdata = HR_data_test)

#our predicted salary is (61131.74)
HR_data_test$monthly.income.of.employee.pred[31]

#since the  predicted salary is almost similar to claimed salary it is said that candidate is genuine 






